source/common/event/timer_impl.h

```diff
@@ -16,7 +16,33 @@ namespace Event {
  */
 class TimerUtils {
 public:
-  static void millisecondsToTimeval(const std::chrono::milliseconds& d, timeval& tv);
+  /**
+   * Intended for consumption by enable(HR)Timer, this method is templated method to avoid implicit
+   * duration conversions for its input arguments. This lets us have an opportunity to check bounds
+   * before doing any conversions. When the passed in duration exceeds UINT32_T max days, the output
+   * will be clipped to yield INT32_MAX days worth of seconds and 0 microseconds for the output
+   * argument. ASSERTS on negative duration input.
+   * @tparam Duration std::chrono duration type, e.g. seconds, milliseconds, ...
+   * @param d duration value
+   * @param tv output parameter that will be updated
+   */
+  template <typename Duration> static void durationToTimeval(const Duration& d, timeval& tv) {
+    if (d.count() < 0) {
```

Comment:
Additional thought: one thing that might be worth attempting here, is to typedef the duration args to `enableXXXTimer()`. E.g.
```c++ 
using TimerDuration = std::chrono::duration<uint64_t, std::milli>;
```

Doing that would void the need for this sanity check here, but would also bloat this PR a lot as call sites would have to be audited / updated.

Reply:
Not sure if it is a big win. Perhaps this should be an ASSERT? Or treat negative duration as 0?